cmd/car/compile.go

```diff
@@ -0,0 +1,455 @@
+			break
+		}
+		rootLine, _, err := br.ReadLine()
+		if err != nil {
+			return err
+		}
+		if strings.HasPrefix(string(rootLine), "root ") {
+			var rCidS string
+			fmt.Sscanf(string(rootLine), "root %s", &rCidS)
+			rCid, err := cid.Parse(rCidS)
+			if err != nil {
+				return err
+			}
+			roots = append(roots, rCid)
+		}
+	}
+
+	//parse blocks.
+	rawBlocks := make(map[cid.Cid][]byte)
+	rawCodecs := make(map[cid.Cid]string)
+
+	for {
+		nextCid, mode, nextBlk, err := parsePatch(br)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		rawBlocks[nextCid] = nextBlk
+		rawCodecs[nextCid] = mode
+	}
+
+	//fmt.Printf("structuring as tree...\n")
+	// structure as a tree
+	childMap := make(map[cid.Cid][]cid.Cid)
+	for c := range rawBlocks {
+		if _, ok := childMap[c]; !ok {
+			childMap[c] = make([]cid.Cid, 0)
+		}
+		for d, blk := range rawBlocks {
+			if c.Equals(d) {
+				continue
+			}
+			if strings.Contains(string(blk), c.String()) {
+				if _, ok := childMap[d]; !ok {
+					childMap[d] = make([]cid.Cid, 0)
+				}
+				childMap[d] = append(childMap[d], c)
+			} else if strings.Contains(string(blk), string(c.Bytes())) {
+				if _, ok := childMap[d]; !ok {
+					childMap[d] = make([]cid.Cid, 0)
+				}
+				childMap[d] = append(childMap[d], c)
+			}
+		}
+	}
+
+	//fmt.Printf("rebuilding...\n")
+	// re-parse/re-build CIDs
+	outBlocks := make(map[cid.Cid][]byte)
+	for len(childMap) > 0 {
+		for origCid, kids := range childMap {
+			if len(kids) == 0 {
+				// compile to final cid
+				blk := rawBlocks[origCid]
+				finalCid, finalBlk, err := serializeBlock(c.Context, origCid.Prefix(), rawCodecs[origCid], blk)
+				if err != nil {
+					return err
+				}
+				outBlocks[finalCid] = finalBlk
+
+				// update other remaining nodes of the new cid.
+				for otherCid, otherKids := range childMap {
+					for i, otherKid := range otherKids {
+						if otherKid.Equals(origCid) {
+							if !finalCid.Equals(origCid) {
+								// update block
+								rawBlocks[otherCid] = bytes.ReplaceAll(rawBlocks[otherCid], origCid.Bytes(), finalCid.Bytes())
+								rawBlocks[otherCid] = bytes.ReplaceAll(rawBlocks[otherCid], []byte(origCid.String()), []byte(finalCid.String()))
+							}
+							// remove from childMap
+							nok := append(otherKids[0:i], otherKids[i+1:]...)
+							childMap[otherCid] = nok
+							break // to next child map entry.
+						}
+					}
+				}
+
+				delete(childMap, origCid)
+			}
+		}
+	}
+
+	if !v2 {
+		// write output
+		outStream := os.Stdout
+		if c.IsSet("output") {
+			outFileName := c.String("output")
+			if outFileName == "" {
+				outFileName = carName
+			}
+			outFile, err := os.Create(outFileName)
+			if err != nil {
+				return err
+			}
+			defer outFile.Close()
+			outStream = outFile
+		}
+
+		if err := carv1.WriteHeader(&carv1.CarHeader{
+			Roots:   roots,
+			Version: 1,
+		}, outStream); err != nil {
+			return err
+		}
+		for c, blk := range outBlocks {
```

Comment:
the failing tests are because of this, and the same block below for v1 -- the blocks are ending up shuffled thanks to iterating over the go `map[]` .. we're going to need to keep a slice of CIDs to iterate over and then rewrite those when you do the reconstruction.